### DISCUSSION TITLE
VAAPI invalid chip id or driver is being used.

### DISCUSSION THREAD

**Discussion by Nycrera** [gstreamer-java/gst1-java-core#255] [General]
Hey, I am trying to get gstreamer to stream an mpegts stream over udp with hardware accelerated encoding on video using VAAPI. With intel integrated GPU.
Basically I can run this command in command line to do this and it works without any problems.

`gst-launch-1.0 ximagesrc ! video/x-raw,framerate=30/1 ! timeoverlay ! videoconvert ! vaapih264enc ! queue ! mpegtsmux name=m ! rtpmp2tpay ! udpsink host=127.0.0.1 port=1234 sync=false alsasrc device=hw:0 ! audioconvert ! fdkaacenc ! m.`

But when I am using gstreamer-java with a simple parselaunch test:
```java
Gst.init(Version.BASELINE);
Pipeline pipeline = (Pipeline) Gst.parseLaunch("ximagesrc ! video/x-raw,framerate=30/1 ! timeoverlay ! videoconvert ! vaapih264enc ! queue ! mpegtsmux name=m ! rtpmp2tpay ! udpsink host=127.0.0.1 port=1234 sync=false alsasrc device=hw:0 ! audioconvert ! fdkaacenc ! m.");
pipeline.play();
```
As you can see both pipelines are identical.
However I am getting an error from gstreamer:
```
DRM_IOCTL_I915_GEM_APERTURE failed: Invalid argument
Assuming 131072kB available aperture size.
May lead to reduced performance or incorrect rendering.
get chip id failed: -1 [22]
param: 4, val: 0
```
Now my idea is that, somehow VAAPI is using the wrong driver while working in JVM but the correct one in command line.
There is this Environment variable that needs to be set for the VAAPI to tell which driver to use:
LIBVA_DRIVER_NAME, which I have set to correct value for my GPU, as, LIBVA_DRIVER_NAME=iHD.
Is there a possibility that when gstreamer is ran through this library, this environment variable is not passed to the gstreamer process?

I can see the environment variable using `System.getenv("LIBVA_DRIVER_NAME")`

OR Is this absolutely unrelated to this?

**Comment by Nycrera** [comment]
Basically what makes me paranoid is the line:
`DRM_IOCTL_I915_GEM_APERTURE failed: Invalid argument`
the "I915" makes me think that the wrong driver is being used. Because the commonly used driver for Intel gpu's seems to be the I915 one. Maybe this is the default?

I have set up my VAAPI following the [https://wiki.archlinux.org/title/Hardware_video_acceleration#Intel](https://wiki.archlinux.org/title/Hardware_video_acceleration#Intel)

But in the end why would it work on command line and not work on the library with a simple parseLaunch as this.

**Comment by Nycrera** [comment]
I have updated to `Gst.init(Version.of(1,20))` No change.

**Comment by neilcsmith-net** [comment]
Moving over to discussions - please use the [mailing list](https://groups.google.com/forum/#!forum/gstreamer-java) ideally, or discussions, if you have usage questions.

The `Version.of()` support in `Gst::init` is not for choosing the version of GStreamer you're loading, but to control what API mappings you can access on the Java side and verify you're using the necessary native version.

Try setting GStreamer debug levels using `GLib::setEnv`, as in the commented out line at https://github.com/gstreamer-java/gst1-java-examples/blob/master/WebRTCSendRecv/src/main/java/org/freedesktop/gstreamer/examples/WebRTCSendRecv.java#L78 see if you get any more useful info.

**Reply by Nycrera** [reply]
This is really odd, when I set `GLib.setEnv("GST_DEBUG", "4", true)` I can not reproduce the issue.
What does Glib.setEnv might be doing to fix this? Is it possible that when I use `GLib.setEnv` that the previously mentioned environment variable LIBVA_DRIVER_NAME is getting set too?

**Reply by neilcsmith-net** [reply]
Try setting `LIBVA_DRIVER_NAME` with `GLib::setEnv` and see if that makes a difference.

Are you using that line directly before Gst::init ?  If so, I can't see how it affects this except possibly that debug output changes something in GStreamer or alters a race condition.

**Reply by Nycrera** [reply]
I am using it before Gst::init, yes. I tried setting LIBVA_DRIVER_NAME like you said and I got the same error.
`GLib.setEnv("LIBVA_DRIVER_NAME", "iHD", true);` Like so.
```
DRM_IOCTL_I915_GEM_APERTURE failed: Invalid argument
Assuming 131072kB available aperture size.
May lead to reduced performance or incorrect rendering.
get chip id failed: -1 [22]
param: 4, val: 0
```
I am just as confused.

**Reply by Nycrera** [reply]
I can see, from the debug logs that correct driver iHD is used when debug is enabled.
```
 GST_CONTEXT gstvaapivideocontext.c:165:context_pad_query:<vaapiencodeh264-0:src>[00m context pad peer query failed
0:00:00.108200439 [31m 5232[00m 0x7f639cc1de00 [36mINFO   [00m [00m         GST_CONTEXT gstvaapivideocontext.c:165:context_pad_query:<vaapiencodeh264-0:sink>[00m context pad peer query failed
0:00:00.108204825 [31m 5232[00m 0x7f639cc1de00 [36mINFO   [00m [00m         GST_CONTEXT gstvaapivideocontext.c:238:_gst_context_query:<vaapiencodeh264-0>[00m posting `need-context' message
0:00:00.109315644 [31m 5232[00m 0x7f639cc1de00 [36mINFO   [00m [00m               vaapi gstvaapiutils.c:95:gst_vaapi_log:[00m VA-API version 1.15.0
0:00:00.109408085 [31m 5232[00m 0x7f639cc1de00 [36mINFO   [00m [00m               vaapi gstvaapiutils.c:95:gst_vaapi_log:[00m User environment variable requested driver 'iHD'
0:00:00.109413462 [31m 5232[00m 0x7f639cc1de00 [36mINFO   [00m [00m               vaapi gstvaapiutils.c:95:gst_vaapi_log:[00m Trying to open /usr/lib/dri/iHD_drv_video.so
0:00:00.109443168 [31m 5232[00m 0x7f639cc1de00 [36mINFO   [00m [00m               vaapi gstvaapiutils.c:95:gst_vaapi_log:[00m Found init function __vaDriverInit_1_15
0:00:00.109952825 [31m 5232[00m 0x7f639cc1de00 [36mINFO   [00m [00m               vaapi gstvaapiutils.c:95:gst_vaapi_log:[00m va_openDriver() returns 0
0:00:00.109960612 [31m 5232[00m 0x7f639cc1de00 [36mINFO   [00m [00m               vaapi gstvaapiutils.c:122:vaapi_initialize:[00m VA-API version 1.15
0:00:00.109965191 [31m 5232[00m 0x7f639cc1de00 [36mINFO   [00m [00m        vaapidisplay gstvaapidisplay.c:976:gst_vaapi_display_create:<vaapidisplayglx0>[00m new display addr=0x7f639cfe8100
```

**Reply by Nycrera** [reply]
I have set the debug level to 2 (WARN) and got the same error with this warning:

```
DRM_IOCTL_I915_GEM_APERTURE failed: Invalid argument
Assuming 131072kB available aperture size.
May lead to reduced performance or incorrect rendering.
get chip id failed: -1 [22]
param: 4, val: 0
0:00:00.043825143 [35m 5836[00m 0x7f3e64fd3470 [33;01mWARN   [00m [00m               vaapi gstvaapiutils.c:78:gst_vaapi_warning:[00m /usr/lib/dri/iHD_drv_video.so init failed
0:00:00.349784915 [35m 5836[00m 0x7f3e64f6c700 [33;01mWARN   [00m [00;35m          aggregator gstaggregator.c:2099:gst_aggregator_query_latency_unlocked:<m>[00m Latency query failed
0:00:00.349828387 [35m 5836[00m 0x7f3e64f6c700 [33;01mWARN   [00m [00;35m          aggregator gstaggregator.c:2099:gst_aggregator_query_latency_unlocked:<m>[00m Latency query failed
```

At least now I know that this is not about the wrong driver being used.

**Reply by Nycrera** [reply]
Just like you guessed this has something to do with timing, a race condition?
As, even with setting the debug level to 4, I get the same error sometimes.

Log result with log level 4, no error:
[https://hastebin.com/enadazasul.apache](https://hastebin.com/enadazasul.apache)
Log result with log level 4, error:
[https://hastebin.com/ubozejabok.apachee](https://hastebin.com/ubozejabok.apach)

This is also probably has something to do with my system having 2 GPUs. I'll try with a different system until the weekend.

**Comment by Nycrera** [comment]
I think I have confirmed this is a problem being caused by having two GPUs, one integrated and a dedicated one.
I can use whichever I want, by editing the loading order of kernel modules for the GPUs on command line.
However, when used with gstreamer-java there seems to be some timing/race condition that breaks this.
I have solved the problem for my use case by disabling the dedicated GPU with optimus-manager.
